rfcs/000X-flux-bootstrap-oci/README.md

```diff
@@ -0,0 +1,396 @@
+command in CI pipelines.
+
+### Goals
+
+- Add support to the Flux CLI for bootstrapping with a container registry as the source of truth.
+- Make it easy for users to switch from Git repositories to OCI repositories.
+
+### Non-Goals
+
+- Automate the migration of Flux manifests from a Git bootstrap repository to OCI.
+
+## Proposal
+
+Implement the `flux bootstrap oci` command with the following specifications:
+
+```shell
+flux bootstrap oci \
+--url=oci://<registry-url>/<flux-manifests>:<tag> \
+--username=<registry-username> \
+--password=<registry-password> \
+--kustomization=<local/path/to/kustomization.yaml> \
+--cluster-url=oci://<registry-url>/<fleet-manifests>:<tag> \
+--cluster-path=<path/inside/oci/artifact>
+```
+
+The bootstrap shares the following flags with the `flux install` command:
+
+```text
+--cluster-domain string      internal cluster domain (default "cluster.local")
+--components strings         list of components, accepts comma-separated values (default [source-controller,kustomize-controller,helm-controller,notification-controller])
+--components-extra strings   list of components in addition to those supplied or defaulted, accepts values such as 'image-reflector-controller,image-automation-controller'
+--image-pull-secret string   Kubernetes secret name used for pulling the toolkit images from a private registry
+--log-level logLevel         log level, available options are: (debug, info, error) (default info)
+--network-policy             deny ingress access to the toolkit controllers from other namespaces using network policies (default true)
+--registry string            container registry where the toolkit images are published (default "ghcr.io/fluxcd")
+--toleration-keys strings    list of toleration keys used to schedule the components pods onto nodes with matching taints
+--version string             toolkit version, when specified the manifests are downloaded from https://github.com/fluxcd/flux2/releases
+--watch-all-namespaces       watch for custom resources in all namespaces, if set to false it will only watch the namespace where the toolkit is installed (default true)
+```
+
+The Terraform/OpenTofu counterpart is the `flux_bootstrap_oci` provider that exposes
+the same configuration options as the CLI.
+
+The bootstrap operations are split into two phases:
+
+- Install and self-update configuration for the Flux components.
+- Cluster state reconciliation configuration.
+
+### Install and self-update configuration
+
+The command performs the following steps based on the `url`, `username`,
+`password` and `kustomization` arguments:
+
+1. Logs in to the OCI registry using the provided credentials.
+2. Generates an OCI artifact from the Flux components manifests and the `kustomization.yaml` file.
+3. Applies the Flux components manifests along with their customisations to the cluster.
+4. Pushes the OCI artifact to the container registry using the specified tag.
```

Comment:
I wonder if it would make sense to have two registry auth. A read-only for the image pull secrets, and read-write for pushing the artifacts to the registry. Storing a read-write secret in the cluster for image pull secrets does not seem like a good idea.

Reply:
It is even possible to consider more pull secrets with specific permissions (least privileges principle):
- one with a read-only permission to be used by the Deployments of the Flux controllers to pull images
- one with a read-only permission (different namespace than images) to be used by the OCIRepository resource to pull OCI artifacts
- one with write permissions to be used by the bootstrap command line

Then, it is possible to consider different OCI registry: one for images and another for Flux artifacts, because the latter could contain sensible infrastructure information.

Reply:
> I wonder if it would make sense to have two registry auth. A read-only for the image pull secrets, and read-write for pushing the artifacts to the registry.

Yes this is something the command could support. Currently our OCI implementation supports reading the Docker config file from the host OS, so we could use that for write operations and the flags for in-cluster secret.

@sestegra the pull secret for the container images is already supported, it's one of the common flags to all bootstrap commands.